### ISSUE TITLE
connection reuse - DoQ, DoT, DoH

### ISSUE THREAD

**Issue by mschirrmeister** [tykling/dns_exporter#114]
Hello,

is there a possibility to do a connection reuse for the secure protocols like DoQ, DoT or DoH?.

When I do queries for example with DoT I see FIN packets after one query when I capture packets.
When I do queries with DoQ, for example like this `http://10.0.3.189:15353/query?server=dns-unfiltered.adguard.com&protocol=doq&query_name=example.com&family=ipv4` I think it is also closing the connection after each query. Because the DNS query times are relative high. The dns_exporter log shows the following when I use DoQ.

```
2024-04-22 23:35:29 +0200 INFO quic.receive_datagram():931:  [0bc937959295c2f5] Retrying with token (86 bytes)
2024-04-22 23:35:29 +0200 INFO quic._handle_crypto_frame():1682:  [0bc937959295c2f5] ALPN negotiated protocol doq
2024-04-22 23:35:29 +0200 INFO quic.datagrams_to_send():562:  [0bc937959295c2f5] Connection close sent (code 0x0, reason )

2024-04-22 23:35:31 +0200 INFO quic._handle_crypto_frame():1686:  [9938daf5338c8433] Duplicate CRYPTO data received for epoch Epoch.HANDSHAKE
2024-04-22 23:35:31 +0200 INFO quic._handle_crypto_frame():1682:  [9938daf5338c8433] ALPN negotiated protocol doq
2024-04-22 23:35:31 +0200 INFO quic.datagrams_to_send():562:  [9938daf5338c8433] Connection close sent (code 0x0, reason )
```

Having to go through the full TLS handshake for each query is a lot of overhead, hence the question is there a way to configure that a connection can be hold open for for an **N** amount of time?
If that is possible, the graphing would show the **pure** DNS reponse time. Right now the time includes everything, tcp connection establishment (if TCP protocol), plus TLS connection establishment plus the final application layer DNS query. 

Based on the numbers, that gives a feeling that a secure connection is _much_ slower compared to pure UDP/TCP. Where in fact for a DNS server who does DoQ/DoT upstream, only the first query is slower. But once the connection is established, all subsequent queries are fast.

**Comment by tykling** [comment]
I agree this would be nice but I fear it wont be trivial. However I will look into it and update this issue. Thanks for the suggestion, it is a good idea.

**Comment by mschirrmeister** [comment]
I can imagine that it is not that easy. Not sure if dnspython has something in it. I look in their repo but I did not find much about connection reuse. If there is something in there, not sure if that could make it easier.

The other dns servers/proxies that I use are go based and they have some good connection handling in their code. But they are not a good fit to do consistent good monitoring for specific target dns servers.

My background is that I use the dns_exporter to monitor some DNS servers. But if it creates a new TLS connection for every query, then the DNS response is very high and you would think the server is slow. 
Es an example the average response time with an established connection is around 20ms vs without it is about 400-500ms.

**Comment by tykling** [comment]
dnspython has nothing currently but I might propose adding functionality to make the methods return the socket they create. But I already have two issues open in dnspython so I will wait a bit, and also investigate a bit more locally.

I totally get the issue, it has been known from day 1, and it is "by design" that each lookup is completely independent. But looking at the code it wouldn't be a huge job to do it so stay tuned :)

**Comment by tykling** [comment]
OK so I took a deep dive yesterday. Introducing connection reuse will get fiddly unless I can convince dnspython to make some changes to make it easier to get the socket object. I am thinking something like dnspython adding get_dot_socket(), get_doh_socket() etc methods that it calls internally like today but which are also easy to call externally.

I had a thought, though. If it would be good enough for you to just measure the query time without the tcp and tls handshake dnspython does actually return this info. It would be a simple matter to use that value instead of timing the whole operation. But the full handshake would have to be measured as well. The question then becomes which of the two metrics get to keep the current name :) I guess it would have to be the existing one (measuring the full handshake) to avoid breaking dashboards.

Would that solve your issue?

**Comment by tykling** [comment]
OK, scratch that, the time dnspython returns also includes the handshake times. Annoying. OK I will ponder this some more.

**Comment by tykling** [comment]
There is a more basic issue thought: dnspython closes the socket right after a query even if the socket was externallyu provided rather than created inside dnspython. This will have to get changed somehow for this idea to have any chance of working, obviously. I have opened https://github.com/rthalley/dnspython/issues/1176 for that, lets see what becomes of that :)

**Comment by mschirrmeister** [comment]
I had the same thought at one point if it would be possible to just measure the actual dns response time internally and expose this.

If this can be done, I think it would be actually a really good thing to have both metrics. If you configure the same server twice, one with connection re-use the other without, you can compare and visualize how much of a difference there is.
